cupy/cutensor.py

```diff
@@ -307,3 +316,76 @@ def contraction_max_algos():
     See cupy/cuda/cutensor.contractionMaxAlgos() for details.
     """
     return cutensor.contractionMaxAlgos()
+
+
+def reduction(alpha, A, desc_A, mode_A, beta, C, desc_C, mode_C,
+              reduce_op=cutensor.OP_ADD, compute_dtype=None):
+    """Tensor reduction
+
+    This routine computes the tensor reduction:
+
+        C = alpha * reduce_op(uop_A(A)) + beta * uop_C(C))
+
+    See cupy/cuda/cutensor.reduction for details.
+
+    Args:
+        alpha: Scaling factor for A.
+        A (cupy.ndarray): Input tensor.
+        desc_A (class Descriptor): A descriptor that holds the information
+            about the data type, modes, strides and unary operator (uop_A) of
+            tensor A.
+        mode_A (tuple of int/str): A tuple that holds the labels of the modes
+            of tensor A (e.g., if A_{x,y,z}, mode_A = {'x','y','z'})
+        beta: Scaling factor for C.
+        C (cupy.ndarray): Input/output tensor.
+        desc_C (class Descriptor): A descriptor that holds the information
+            about the data type, modes, strides and unary operator (uop_C) of
+            tensor C.
+        mode_C (tuple of int/str): A tuple that holds the labels of the modes
+            of tensor C.
+        reduce_op (cutensorOperator_t): Binary operator used to reduce A.
+        compute_dtype (numpy.dtype): Compute type for the intermediate
+            computation.
+
+    Returns:
+        out (cupy.ndarray): Output tensor.
+
+    Examples:
+        See examples/cutensor/reduction.py
+    """
+    assert A.dtype == C.dtype
+    assert A.ndim == len(mode_A)
+    assert C.ndim == len(mode_C)
+    mode_A = _convert_mode(mode_A)
+    mode_C = _convert_mode(mode_C)
+    out = C
+    if compute_dtype is None:
+        if A.dtype == numpy.float16:
+            compute_dtype = numpy.float32
+        else:
+            compute_dtype = A.dtype
+    alpha = numpy.array(alpha, compute_dtype)
+    beta = numpy.array(beta, compute_dtype)
+    handle = get_handle()
+    compute_dtype = get_cuda_dtype(compute_dtype)
+    ws_size = cutensor.reductionGetWorkspace(
+        handle,
+        A.data.ptr, desc_A.value, mode_A.ctypes.data,
+        C.data.ptr, desc_C.value, mode_C.ctypes.data,
+        out.data.ptr, desc_C.value, mode_C.ctypes.data,
+        reduce_op, compute_dtype)
+    try:
+        ws = cupy.ndarray((ws_size,), dtype=numpy.int8)
+    except Exception:
```

Comment:
Will cuTensor without workspace? In that case, could you catch `cupy.cuda.memory.OutOfMemoryError` as other errors are unlike to recover?